### ISSUE TITLE
Cannot call "initialize" more than once

### ISSUE THREAD

**Issue by edward-hong** [evanw/esbuild#1266]
I'm creating multiple code editors in React but it throws the error `Cannot call "initialize" more than once` when I try to initialize multiple times. According to the docs:

```
If you're already running this code from a worker and don't want initialize to create another worker, you can pass worker: false to it. Then it will create a WebAssembly module in the same thread as the thread that calls initialize.
```

So the first time I initialize, I set the `worker` property to `true` and then with subsequent initializations, I call it with the `worker` property to `false`. However I'm still getting the same error. Any help with this issue is appreciated.

**Comment by evanw** [comment]
I'm confused. The error is correct. The library is not designed to be initialized more than once within a given JS VM instance. You can initialize the worker either to run its code on the same thread (`worker: false`) or to generate its own worker thread and run its code there (`worker: true`), but you can pick one of those options. Once you call `initialize()` then the library will have started to initialize, and doesn't support being initialized again from that point onward.

The point of `worker: false` is that if you are already in a worker thread of your own creation (`main thread -> your worker thread`), you may not want esbuild to create its own worker thread (`main thread -> your worker thread -> esbuild's worker thread`) since that could be wasteful of CPU and RAM if your worker thread is going to be otherwise idle the whole time. That option is just an optimization and doesn't need to be used if you don't need it. You don't need to create a worker thread yourself since esbuild does that internally (in case that's what you are trying to do).

**Comment by evanw** [comment]
Closing this issue since this is working as intended. You are only supposed to call `initialize` once. Details are in the previous comment.

**Comment by swport** [comment]
If you're using React in strict mode then your useEffect, wherein you might be initializing esbuild, is actually running twice. Hence this error. Try removing React.StrictMode.

**Comment by ChristopherHButler** [comment]
I am having a similar issue with to this. I have a nextjs app and I am struggling to find a way to use esbuild without initialize being called more than once.

There is a Udemy course where Stephen Grider creates a code editor of sorts and shows how to use esbuild to transpile and bundle code. When he made the course, it used an older version of esbuild so you could write something like this: 

```js
import * as esbuild from 'esbuild-wasm';

import { unpkgPathPlugin } from './plugins/unpkg-path-plugin';
import { fetchPlugin } from './plugins/fetch-plugin';

let service;

const bundle = async ({ userCode }) => {
  if (!service) {
    service = await esbuild.startService({
      worker: true,
      wasmURL: 'https://unpkg.com/esbuild-wasm@0.8.36/esbuild.wasm',
    });
  }

  const buildOptions = {
    entryPoints: ['index.js'],
    bundle: true,
    write: false,
    plugins: [unpkgPathPlugin(), fetchPlugin(userCode)],
    define: {
      'process.env.NODE_ENV': '"production"',
      global: 'window',
    },
  };

  const result = await service.build({
    ...buildOptions,
  });

  return {
    code: result.outputFiles[0].text,
    err: '',
  };
};

export default bundle;
```

That worked really well, however, since there is no service and `esbuild.initialize` does not return any value, I'm not sure how I can guarantee it does not get called twice.  I've looked for answers on stack overflow and found these: 

- [https://stackoverflow.com/questions/66563719/esbuild-startservice-replacement](https://stackoverflow.com/questions/66563719/esbuild-startservice-replacement)
- [https://stackoverflow.com/questions/66563719/esbuild-startservice-replacement](https://stackoverflow.com/questions/66563719/esbuild-startservice-replacement)

Unfortunately, none of the responses here work for me. I've also found [this article](https://kristiansigston.medium.com/initializing-esbuild-wasm-84a26b405f12) where someone has explained how they have got a newer version of esbuild working for this purpose but, again, I am unable to make it work. 

I have written the above bundle function outside of a react component and it is not a hook. I do call this function from insode of a react `useEffect` many times to rebundle the code any time a user makes a change in the editor. 

My current code looks like this: 

```ts
import * as esbuild from 'esbuild-wasm';
import { unpkgPathPlugin } from './plugins/unpkg-path-plugin';
import { fetchPlugin } from './plugins/fetch-plugin';
import { Experiment } from '@/src/types';
import { getExperimentEntryFile } from '@/src/utils/experimentUtils';



interface BundleProps {
  experiment: Experiment;
}

export const bundle = async ({ experiment }: BundleProps) => {

  // need this in case esbuild is not initialized yet.
  await esbuild.initialize({
    // create it's own worker thread
    worker: true,
    // Note: IF YOU UPDATE THE VERSION OF ESBUILD IN PACKAGE.JSON YOU MUST UPDATE THIS AS WELL
    wasmURL: 'https://unpkg.com/esbuild-wasm@0.21.4/esbuild.wasm',
    
  });

  // get the entryFile
  const entryFile = getExperimentEntryFile({ experiment });

  // plugin order matters here
  const plugins = [
    unpkgPathPlugin(),
    fetchPlugin({ entryFile, experiment }),
  ];

  const buildOptions = {
    entryPoints: [entryFile.name],
    bundle: true,
    write: false,
    plugins,
    define: {
      'process.env.NODE_ENV': '"production"', // use the production version of packages from unpkg
      global: 'window'                                            // webpack does this. Some packages will refer to global and it won't work.
    },
  }

  const result = await esbuild.build({
    ...buildOptions,
  });

  return {
    bundledCode: result.outputFiles[0].text,
  };

};

```

Again, everything works fine calling bundle the first time, but I get an error on the second call (obviously) and have no idea how to fix it. Any help here would be greatly appreciated! Also, since I did not start the thread I am unsure what other information might be required but I am happy to help in any way I can to get it working together!

**Comment by evanw** [comment]
That's really more of a general question about how to do asynchronous JavaScript, not about how esbuild's API works. Here's one way to do it (how to call a JavaScript function once but await its returned promise every time):

```patch
-let service;
+let initialize;
 
 const bundle = async ({ userCode }) => {
-  if (!service) {
-    service = await esbuild.startService({
+  if (!initialize) {
+    initialize = esbuild.initialize({
       worker: true,
       wasmURL: 'https://unpkg.com/esbuild-wasm@0.8.36/esbuild.wasm',
     });
   }
+  await initialize;
 };
```

It's very similar to what you were doing before. Your problem is that you deleted the if statement, which was the thing that made sure to only initialize the library once.

**Comment by ChristopherHButler** [comment]
Thanks @evanw that seems to work!
As mentioned I had tried the following (along with a few other implementations) but i kept getting the same error. 

```ts
let initialize = false;

  if (!initialize) {
     await esbuild.initialize({
      worker: true,
      wasmURL: 'https://unpkg.com/esbuild-wasm@0.21.4/esbuild.wasm',
    });
    initialize = true;
  }
```
Anyways, I appreciate the help and I'm sure it will help some others as well.

**Comment by ngocla99** [comment]
@ChristopherHButler ,
Amazing! I'm studying this course now and have the same problem as you. I'm using the version esbuild-wasm: 0.21.1.

At first, I tried to solve the problem by defining a global variable and modifying it when the esbuild initialized. Then, I would check if it was modified to avoid running esbuild.initialize() again. This is similar to using the singleton pattern.

However, I realized that when the app runs, multiple bundles may be called at the same time, so the global variable might not be modified yet (perhaps due to the time it takes to call esbuild.initialize()).

So, I decided to initialize it in the App

`//index.tsx
import "bulmaswatch/superhero/bulmaswatch.min.css";
import ReactDOM from "react-dom/client";
import { Provider } from "react-redux";
import { CellList } from "./components";
import { store } from "./state";
import React from "react";
import startEsbuild from "./bundler/startEsbuild";

`const App = () => {
  const [isLoaing, setIsLoading] = React.useState(true);

  React.useEffect(() => {
    (async () => {
      try {
        await startEsbuild();
        setIsLoading(false);
      } catch (err) {
        setIsLoading(false);
      }
    })();
  }, []);

  if (isLoaing) return <div>Loading...</div>;

  return (
    <Provider store={store}>
      <div className="">
        <CellList />
      </div>
    </Provider>
  );
};

const root = ReactDOM.createRoot(
  document.getElementById("root") as HTMLElement
);

root.render(<App />);
`
